src/xform_to_datalog/datalog_facts_test.cc

```diff
@@ -0,0 +1,492 @@
+//-----------------------------------------------------------------------------
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//-----------------------------------------------------------------------------
+
+#include "src/xform_to_datalog/datalog_facts.h"
+
+#include <google/protobuf/text_format.h>
+
+#include "src/common/testing/gtest.h"
+
+namespace raksha::xform_to_datalog {
+
+namespace ir = raksha::ir;
+
+class DatalogFactsToStringTest :
+   public testing::TestWithParam<std::tuple<DatalogFacts, std::string>> {};
+
+TEST_P(DatalogFactsToStringTest, DatalogFactsToStringTest) {
+  const DatalogFacts &datalog_facts = std::get<0>(GetParam());
+  const std::string &expected_result_string = std::get<1>(GetParam());
+
+  EXPECT_EQ(datalog_facts.ToString(), expected_result_string);
+}
+
+static const ir::AccessPath kHandleH1AccessPath(
+    ir::AccessPathRoot(ir::HandleAccessPathRoot("recipe", "h1")),
+    ir::AccessPathSelectors());
+
+static const ir::AccessPath kHandleH2AccessPath(
+    ir::AccessPathRoot(ir::HandleAccessPathRoot("recipe", "h2")),
+    ir::AccessPathSelectors());
+
+static const ir::AccessPath kHandleConnectionInAccessPath(
+    ir::AccessPathRoot(ir::HandleConnectionAccessPathRoot(
+        "recipe", "particle", "in")),
+    ir::AccessPathSelectors());
+
+static const ir::AccessPath kHandleConnectionOutAccessPath(
+    ir::AccessPathRoot(ir::HandleConnectionAccessPathRoot(
+        "recipe", "particle", "out")),
+    ir::AccessPathSelectors());
+
+static std::tuple<DatalogFacts, std::string>
+  datalog_facts_and_output_strings[] = {
+    { DatalogFacts({}, {}, {}), "// Claims:\n\n// Checks:\n\n// Edges:\n" },
+    { DatalogFacts(
+        { ir::TagClaim(ir::TagAnnotationOnAccessPath(
+            kHandleConnectionOutAccessPath, "tag")) },
+        { ir::TagCheck(ir::TagAnnotationOnAccessPath(
+            kHandleConnectionInAccessPath, "tag2")) },
+        { ir::Edge(kHandleH1AccessPath, kHandleConnectionInAccessPath),
+          ir::Edge(kHandleConnectionInAccessPath,
+                   kHandleConnectionOutAccessPath),
+           ir::Edge(kHandleConnectionOutAccessPath, kHandleH2AccessPath)}),
+      "// Claims:\nclaimHasTag(\"recipe.particle.out\", \"tag\").\n"
```

Comment:
Use rawstrings for better readability of the tests: https://abseil.io/tips/64

Also, can you file a bug to move all the other tests to use raw strings wherever possible as well?

Reply:
Bug filed as: https://github.com/google-research/raksha/issues/109

Reply:
Done.